include/ranges.h

```diff
@@ -64,6 +66,17 @@ struct subrange {
 	subrange(chunk<Dims> other) : offset(other.offset), range(other.range) {}
 
 	bool operator==(const subrange& rhs) { return offset == rhs.offset && range == rhs.range; }
+
+	std::string to_string() const {
+		if(dims == 1)
+			return fmt::format("subrange({}) at offset({})", range[0], offset[0]);
+		else if(dims == 2)
+			return fmt::format("subrange({},{}) at offset({},{})", range[0], range[1], offset[0], offset[1]);
+		else
+			return fmt::format("subrange({},{},{}) at offset({},{},{})", range[0], range[1], range[2], offset[0], offset[1], offset[2]);
+	}
+
+	size_t num_elems() const { return range.size(); }
```

Comment:
The subrange is a generic "range with an offset", so I think its somewhat misleading to refer to "elements" here. After all, the subrange doesn't always describe the subset of a buffer, but instead could also describe the number of "items" in a kernel execution space.

That being said, `subrange.range` is accessible anyways, so why don't you just use `subrange.range.size()` for this?